maintenance/wikia/Forum/fixCommentIndexes.php

```diff
@@ -0,0 +1,262 @@
+
+		foreach ( $threads as $threadId ) {
+			$brokenThread = new BrokenThread( $threadId );
+			$brokenThread->fix();
+		}
+	}
+
+	/**
+	 * Find the list of threads with children that have duplicated values in the props field
+	 * of page_wikia_props.  This indicates the counter has been reused by more than one comment
+	 * in the same thread.  We ignore deleted, removed and archived comments since those are likely
+	 * to have duplicates with current comments because the author of that code never considered
+	 * un-deleting/removing/archiving as a thing.  Not going to fix that here since who knows what
+	 * additional bugs that will uncover.
+	 *
+	 * @return array
+	 */
+	private function getAffectedThreads() {
+		$dbr = wfGetDB( DB_SLAVE );
+		$threads = ( new WikiaSQL )
+			->SELECT( 'parent_comment_id', 'props', 'count(*)' )
+			->FROM( 'comments_index' )
+			->JOIN( 'page_wikia_props' )
+			->WHERE( 'page_id' )->EQUAL_TO_FIELD( 'comment_id' )
+			->AND_( 'propname' )->EQUAL_TO( WPP_WALL_COUNT )
+			->AND_( 'parent_comment_id' )->NOT_EQUAL_TO( 0 )
+			->AND_( 'removed' )->EQUAL_TO( 0 )
+			->AND_( 'deleted' )->EQUAL_TO( 0 )
+			->AND_( 'archived' )->EQUAL_TO( 0 )
+			->GROUP_BY( 'parent_comment_id, props' )
+			->HAVING( 'count(*)' )->GREATER_THAN( 1 )
+			->runLoop( $dbr, function ( &$threads, $row ) {
+				$threads[] = $row->parent_comment_id;
+			} );
+
+		// Ensure that we're always sending an array back
+		return empty( $threads ) ? [ ] : $threads;
+	}
+
+	/**
+	 * Get the URL for the wiki being worked on by this script
+	 *
+	 * @return string
+	 */
+	static public function getWikiURL() {
+		$dbName = WikiFactory::IDtoDB( F::app()->wg->CityId );
+		$url = WikiFactory::DBtoDomain( $dbName );
+		return WikiFactory::getLocalEnvURL( 'http://'.$url );
+	}
+}
+
+class BrokenThread {
+
+	private $threadId;
+
+	/**
+	 * @param int $threadId
+	 */
+	public function __construct( $threadId ) {
+		$this->threadId = $threadId;
+	}
+
+	/**
+	 * Fix the comment numbering for the thread given by $this->threadId
+	 */
+	public function fix() {
+		FixCommentIndexes::debug( "* Fixing thread ".$this->threadId );
+
+		$lastIdx = $this->renumberComments();
+		$this->setNewWallCount( $lastIdx );
+	}
+
+	/**
+	 * Sets the new index to use whenever the next comment is made on the current thread
+	 *
+	 * @param int $idx Update the new last used index on this wall
+	 */
+	private function setNewWallCount( $idx ) {
+		if ( FixCommentIndexes::isTest() ) {
+			return;
+		}
+
+		wfSetWikiaPageProp( WPP_WALL_COUNT, $this->threadId, $idx );
+	}
+
+	/**
+	 * Update the comment number used for creating links to those comments.  It doesn't matter what they are
+	 * as long as each comment has a unique number, however ascending from 2 is how it works now and makes the
+	 * most sense.
+	 *
+	 * @return int The last index we assigned to a comment
+	 */
+	private function renumberComments() {
+		$comments = $this->getCommentIds();
```

Comment:
That's an understood risk.  I'm trying to keep this as simple as possible to spend the minimum amount of time necessary for a one time fix of a product we'll be replacing in the next few months.

Fetching new comments and checking for new comments don't eliminate the race condition, just moves it around (there is still a nonatomic read followed by a write).  Additionally, it would add complexity to the script and will slow execution considerably.

Locking is the only way to ensure nothing updates between determining the value to use for the wall count, and setting that value.  However I would have obtain a lock on `DB_MASTER` and use that handle for both reads and writes.  This means this script would be hammering the master DB across all clusters while running and would repeatedly lock the table used for all commenting on a wiki, not just forums.

As a compromise I can reduce the race condition window.  If I did the `count(*)` on the comment index table followed immediately with `wfSetWikiaPageProp(WPP_WALL_COUNT, ... )` for the thread, there would be a much smaller window for new comments to be made.  Any replies made for the rest of the script execution will use the correct index as we renumber all replies behind it.  A reply made between the `count(*)` read and the `WPP_WALL_COUNT` write would create a repeated index, the problem this script fixes.  If this happens and if someone complains about it, we can easily re-run the script for that single thread.